demos/github-demo/src/isograph-components/PullRequestLink.tsx

```diff
@@ -11,6 +12,10 @@ export const PullRequestLink = iso(`
       name
       owner {
         login
+        asUser {
```

Comment:
Amazing!!! Thanks for continuing to make progress.

It took me a while to figure out how we should handle this. In particular, I want to make sure what we're building fits into a coherent bigger picture.

## `asUser` vs other types of fields

How does `asUser` differ from `pet(id: 4)`? Both of them can be thought of as client pointers. A client pointer has a `condition` function associated with it. `pet(id: '4')` has a simple condition: `(data: {}) => { __typename: 'Pet', id: '4' }`, and `asUser` has condition `(data: { __typename, id }) => data.__typename === 'User' ? { __typename: 'User', id: data.id } : null`. We aren't actually generating the `pet` function, but that's the implicit behavior.

So, the way I'm thinking about it is that linked fields and pointers are the same thing, except linked fields have a "default" condition.

Okay, with that said, how does `asUser` differ from a proper client pointer. Well, it differs in that it doesn't have a refetch query or entrypoint. The reader AST is sufficient. And since it doesn't have a refetch query or entrypoint, we can just read the result as part of the parent fragment.

## categorizing linked fields

So, for linked fields on the reader side, I see four possible cases:

```
                     no query/entrypoint     has query/entrypoint
unconditional (1)    regular linked field    loadably selected linked field
  conditional        asUser field (2)        client pointer, loadably selected inline fragment
```
- Note that this grid excludes loadably selected client fields and refetch fields, which are scalars!
- (1): the unconditional row just means "default" behavior. You can't have an unconditional client pointer, because we're defining a new field, and so there isn't a link in the store (at that same slot) to use for the default behavior. So, a client pointer which always returns a DataID is still unconditional, in this world.
- (2): also, client pointers that are executed on the server. That's pretty much what `asUser` fields are, with the added twist that GraphQL understands them.

So, I think we can model reader linked fields in all four quadrants as follows:

```diff
export type ReaderLinkedField = {
  readonly kind: 'Linked';
  readonly fieldName: string;
  readonly alias: string | null;
  readonly selections: ReaderAst<unknown>;
  readonly arguments: Arguments | null;
+  readonly condition: LinkedFieldCondition | null // (1)
+  // let's not implement this yet:
+  readonly entrypointOrRefetchQueryIndexOrSomething: Something | null
};
```

Okay, so what is this `LinkedFieldCondition`? I think that this is a generated file, `Node/asUser/resolver_reader.ts`, which we generate for all encountered client pointers and inline fragments. For a client pointer, the condition would be an import of the resolver (i.e. `export const bestFriendClientPointer = ...`). For inline fragment fields, we can hard code the resolver.

An eager reader artifact works as a `LinkedFieldCondition`, and maybe we can reuse that, with the hard coded resolver, and with `TClientFieldValue` (aka the output type) being `Link`.

So, given that, we should actually just have

```diff
+  readonly condition: EagerReaderArtifact<any, Link>
```

(Maybe we can improve upon the `any` type, but I don't think it's important.)

------------

Okay, so that's how the reader AST stuff will have to change. How can we actually do this?

- the schema server field variant `InlineFragment` should have a reader AST containing `__typename` and `id`
    - that should be enough. I think we can just hard code the `({ data }) => data.__typename === ConcreteType ? data.id : null` wherever we generate that for now.
- when generating a merged selection set, that reader AST w/typename and id should additionally be merged into the merged selection set.
- when generating artifacts
    - when we encounter a linked field with an inline fragment, that should be tracked in a data structure similar to `encountered_client_field_map`
    - for each linked field in this data structure, we should generate a resolver reader. That is probably enough. But, the resolver reader references the `param_type.ts` and `output_type.ts`, so maybe it will be easier to just generate those as well. (The output_type.ts can just be `export type Type__fieldName__output_type = Link`, `param_type.ts` will probably be correctly generated, but also doesn't matter.)
    - whenever we encounter an inline fragment, we need to import the generated file, i.e. `PullRequestLink/resolver_resolver` should import `User/asUser/resolver_reader`, just as if it was a client field
- read.ts, when we encounter a linked field, should either read it from the store using `storeRecord[getParentRecordKey(...)]` or by reading `condition.readerAst` and calling `condition.resolver` with that data


Okay, I know that's a lot! We can chat tomorrow if you want about this.